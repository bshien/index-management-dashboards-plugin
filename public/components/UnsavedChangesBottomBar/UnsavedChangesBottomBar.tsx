/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */
import React, { useCallback, useEffect, useRef, useState } from "react";
import { EuiButton, EuiFlexGroup, EuiFlexItem, EuiSmallButtonEmpty, EuiButtonProps, EuiButtonEmptyProps } from "@elastic/eui";
import classNames from "classnames";
import BottomBar from "../BottomBar";
import "./index.scss";
import { createPortal } from "react-dom";

export type CustomFormRowProps = {
  unsavedCount: number;
  formErrorsCount?: number;
  onClickCancel?: () => void;
  onClickSubmit: () => Promise<void>;
  submitButtonDataTestSubj?: string;
  renderProps?: (props: {
    renderCancel: () => React.ReactChild;
    renderConfirm: () => React.ReactChild;
    renderUnsavedText: () => React.ReactChild;
    loading?: boolean;
  }) => React.ReactChild;
  confirmButtonProps?: EuiButtonProps;
  cancelButtonprops?: EuiButtonEmptyProps;
};

export default function UnsavedChangesBottomBar(props: CustomFormRowProps) {
  const { unsavedCount, onClickCancel, onClickSubmit, submitButtonDataTestSubj, formErrorsCount } = props;
  const [loading, setLoading] = useState(false);
  const destroyRef = useRef(false);
  const onClick = async () => {
    setLoading(true);
    try {
      await onClickSubmit();
    } catch (e) {
    } finally {
      if (destroyRef.current) {
        return;
      }
      setLoading(false);
    }
  };

  useEffect(() => {
    return () => {
      destroyRef.current = true;
    };
  }, []);

  const renderCancel = useCallback(
    () => (
      <EuiFlexItem grow={false}>
        <EuiSmallButtonEmpty onClick={onClickCancel} color="ghost" iconType="cross" children="Cancel" {...props.cancelButtonprops} />
      </EuiFlexItem>
    ),
    [onClickCancel]
  );

  const renderConfirm = useCallback(
    () => (
      <EuiFlexItem grow={false}>
        <EuiButton
          data-test-subj={submitButtonDataTestSubj}
          onClick={onClick}
          isLoading={loading}
          disabled={loading}
          iconType="check"
          color="primary"
          fill
          size="s"
          children="Save"
          {...props.confirmButtonProps}
        />
      </EuiFlexItem>
    ),
    [onClick, submitButtonDataTestSubj, loading]
  );

  const renderUnsavedText = useCallback(
    () => (
      <>
        {formErrorsCount ? (
          <EuiFlexItem style={{ flexDirection: "row" }}>
            <div
              className={classNames({
                "ISM-unsaved-changes-blocks": true,
                danger: true,
              })}
            />
            {formErrorsCount} form errors.
          </EuiFlexItem>
        ) : null}
        {unsavedCount && !formErrorsCount ? (
          <EuiFlexItem style={{ flexDirection: "row" }}>
            <div
              className={classNames({
                "ISM-unsaved-changes-blocks": true,
                warning: true,
              })}
            />
            {unsavedCount} unsaved changes.
          </EuiFlexItem>
        ) : null}
      </>
    ),
    [unsavedCount, formErrorsCount]
  );

  const renderProps =
    props.renderProps ||
    (() => (
      <>
        {renderUnsavedText()}
        {renderCancel()}
        {renderConfirm()}
      </>
    ));

  const bottomBar = (
    <BottomBar>
      <EuiFlexGroup alignItems="center" gutterSize="s" justifyContent="spaceBetween">
        {renderProps({
          renderCancel,
          renderConfirm,
          renderUnsavedText,
          loading,
        })}
      </EuiFlexGroup>
    </BottomBar>
  );

  return createPortal(bottomBar, document.getElementById("app-wrapper")!);
}
