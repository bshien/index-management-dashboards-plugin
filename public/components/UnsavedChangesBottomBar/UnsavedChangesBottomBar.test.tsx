/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import { render, cleanup } from "@testing-library/react";
import UnsavedChangesBottomBar from "./index";

describe("<FormGenerator /> spec", () => {
  let appWrapper: HTMLDivElement;

  beforeEach(() => {
    // Create a mock DOM element with the ID 'app-wrapper'
    appWrapper = document.createElement("div");
    appWrapper.setAttribute("id", "app-wrapper");
    document.body.appendChild(appWrapper);
  });

  afterEach(() => {
    // Clean up the mock DOM element after each test
    document.body.removeChild(appWrapper);
    cleanup();
  });

  it("render the component", () => {
    render(<UnsavedChangesBottomBar unsavedCount={0} onClickCancel={async () => {}} onClickSubmit={async () => {}} />);
    expect(document.body.children).toMatchSnapshot();
  });
});
