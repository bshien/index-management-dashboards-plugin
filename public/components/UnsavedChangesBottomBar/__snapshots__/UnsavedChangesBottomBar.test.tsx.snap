// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormGenerator /> spec render the component 1`] = `
HTMLCollection [
  <div
    id="app-wrapper"
  >
    <section
      aria-label="Page level controls"
      class="euiBottomBar euiBottomBar--sticky euiBottomBar--paddingMedium"
    >
      <h2
        class="euiScreenReaderOnly"
      >
        Page level controls
      </h2>
      <div
        class="euiFlexGroup euiFlexGroup--gutterSmall euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      >
        <div
          class="euiFlexItem euiFlexItem--flexGrowZero"
        >
          <button
            class="euiButtonEmpty euiButtonEmpty--ghost euiButtonEmpty--small"
            type="button"
          >
            <span
              class="euiButtonContent euiButtonEmpty__content"
            >
              EuiIconMock
              <span
                class="euiButtonEmpty__text"
              >
                Cancel
              </span>
            </span>
          </button>
        </div>
        <div
          class="euiFlexItem euiFlexItem--flexGrowZero"
        >
          <button
            class="euiButton euiButton--primary euiButton--small euiButton--fill"
            type="button"
          >
            <span
              class="euiButtonContent euiButton__content"
            >
              EuiIconMock
              <span
                class="euiButton__text"
              >
                Save
              </span>
            </span>
          </button>
        </div>
      </div>
    </section>
    <p
      aria-live="assertive"
      class="euiScreenReaderOnly"
    >
      There is a new region landmark with page level controls at the end of the document.
    </p>
  </div>,
  <div />,
]
`;
